File: i3rd/UI/GuideDetailViewController.m

#import "GuideDetailViewController.h"

NSString *const GuideDetailViewControllerCharacterCodeKey = @"CharacterCode";
NSString *const GuideDetailViewControllerSkillCodeKey = @"SkillCode";
NSString *const GuideDetailViewControllerSkillNameKey = @"SkillName";
NSString *const GuideDetailViewControllerViewControllerKey = @"ViewController";
NSString *const GuideDetailViewControllerPresentSkillMotionPlayerViewControllerSegue = @"PresentSkillMotionPlayerViewController";

@interface GuideDetailViewController ()

@end

@implementation GuideDetailViewController

#pragma mark - Lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    
//    UIView *tableViewBackgroundView = [[UIView alloc] initWithFrame:self.tableView.bounds];
//    tableViewBackgroundView.backgroundColor = [UIColor clearColor];
//    self.tableView.backgroundView = tableViewBackgroundView;
//    self.tableView.backgroundColor = [UIColor clearColor];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:GuideDetailViewControllerPresentSkillMotionPlayerViewControllerSegue]) {
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        NSDictionary *sectionInfo = [self.sections objectAtIndex:indexPath.section];
        NSArray *rows = [sectionInfo objectForKey:PropertyListBasedViewControllerRowsKey];
        NSDictionary *rowInfo = [rows objectAtIndex:indexPath.row];
        NSDictionary *presented = [rowInfo objectForKey:PropertyListBasedViewControllerPresentedKey];
        
        SkillMotionPlayerViewController *player = (SkillMotionPlayerViewController *)[segue.destinationViewController topViewController];
        player.delegate = self;
        player.characterCode = [presented objectForKey:GuideDetailViewControllerCharacterCodeKey];
        player.skillCode = [presented objectForKey:GuideDetailViewControllerSkillCodeKey];
        player.title = [presented objectForKey:GuideDetailViewControllerSkillNameKey];
    }
}

#pragma mark - State Restoration

#define GuideDetailViewControllerTitleKey @"Title"
#define GuideDetailViewControllerSelectedIndexPathKey @"SelectedIndexPath"

- (void)encodeRestorableStateWithCoder:(NSCoder *)coder
{
    [super encodeRestorableStateWithCoder:coder];
    
    [coder encodeObject:self.title forKey:GuideDetailViewControllerTitleKey];
    [coder encodeObject:self.sections forKey:PropertyListBasedViewControllerSectionsKey];
    [coder encodeObject:[self.tableView indexPathForSelectedRow] forKey:GuideDetailViewControllerSelectedIndexPathKey];
}

- (void)decodeRestorableStateWithCoder:(NSCoder *)coder
{
    [super decodeRestorableStateWithCoder:coder];
    
    self.title = [coder decodeObjectForKey:GuideDetailViewControllerTitleKey];
    self.sections = [coder decodeObjectForKey:PropertyListBasedViewControllerSectionsKey];
    NSIndexPath *selectedIndexPath = [coder decodeObjectForKey:GuideDetailViewControllerSelectedIndexPathKey];
    [self.tableView selectRowAtIndexPath:selectedIndexPath animated:NO scrollPosition:UITableViewScrollPositionTop];
}

#pragma mark - Table View Delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSDictionary *sectionInfo = [self.sections objectAtIndex:indexPath.section];
    NSArray *rows = [sectionInfo objectForKey:PropertyListBasedViewControllerRowsKey];
    NSDictionary *rowInfo = [rows objectAtIndex:indexPath.row];
    NSString *rowTitle = [rowInfo objectForKey:PropertyListBasedViewControllerRowTitleKey];
    id next = [rowInfo objectForKey:PropertyListBasedViewControllerNextKey];
    NSDictionary *presented = [rowInfo objectForKey:PropertyListBasedViewControllerPresentedKey];
    
    if (next) {
        NSDictionary *propertyListInfo;
        if ([next isKindOfClass:[NSString class]]) {
            NSURL *propertyListURL = [[[NSBundle mainBundle] bundleURL] URLByAppendingPathComponent:next];
            propertyListInfo = [[NSDictionary alloc] initWithContentsOfURL:propertyListURL];
        } else {
            propertyListInfo = next;
        }
        NSArray *nextSectionsInfo = [propertyListInfo objectForKey:PropertyListBasedViewControllerSectionsKey];
        
        GuideDetailViewController *nextViewController = [self.storyboard instantiateViewControllerWithIdentifier:@"GuideDetailViewController"];
        nextViewController.title = rowTitle;
        nextViewController.sections = nextSectionsInfo;
        [self.navigationController pushViewController:nextViewController animated:YES];
    } else if (presented) {
        NSString *presentedViewControllerName = [presented objectForKey:GuideDetailViewControllerViewControllerKey];
        if ([presentedViewControllerName isEqualToString:@"FramesPlayerViewController"]) {
            [self performSegueWithIdentifier:GuideDetailViewControllerPresentSkillMotionPlayerViewControllerSegue sender:self];
        }
    }
}

#pragma mark - Skill Motion Player View Controller Delegate

- (void)willDismissSkillMotionPlayerViewController:(SkillMotionPlayerViewController *)skillMotionPlayerViewController
{
    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad) {
        [self.tableView deselectRowAtIndexPath:[self.tableView indexPathForSelectedRow] animated:YES];
    }
}

@end

File: i3rd/UI/GuideDetailViewController.swift
//
//  GuideDetailViewController.swift
//  i3rd
//
//  Created by Leon Li on 2018/6/14.
//  Copyright © 2018 Leon & Vane. All rights reserved.
//

import UIKit

let GuideDetailViewControllerCharacterCodeKey = "CharacterCode"
let GuideDetailViewControllerSkillCodeKey = "SkillCode"
let GuideDetailViewControllerSkillNameKey = "SkillName"
let GuideDetailViewControllerViewControllerKey = "ViewController"
let GuideDetailViewControllerPresentSkillMotionPlayerViewControllerSegue = "PresentSkillMotionPlayerViewController"

class GuideDetailViewController: PropertyListBasedViewController {
    override func viewDidLoad() {
        super.viewDidLoad()
        
//        let tableViewBackgroundView = UIView(frame: tableView.bounds)
//        tableViewBackgroundView.backgroundColor = .clear
//        tableView.backgroundView = tableViewBackgroundView
//        tableView.backgroundColor = .clear
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == GuideDetailViewControllerPresentSkillMotionPlayerViewControllerSegue {
            let indexPath = tableView.indexPathForSelectedRow!
            let sectionInfo = sections[indexPath.section] as? NSDictionary
            let rows = sectionInfo?[PropertyListBasedViewControllerRowsKey] as? NSArray
            let rowInfo = rows?[indexPath.row] as? NSDictionary
            let presented = rowInfo?[PropertyListBasedViewControllerPresentedKey] as? NSDictionary
            
            let player = (segue.destination as! UINavigationController).topViewController as! SkillMotionPlayerViewController
            player.delegate = self
            player.characterCode = presented?[GuideDetailViewControllerCharacterCodeKey] as? String
            player.skillCode = presented?[GuideDetailViewControllerSkillCodeKey] as? String
            player.title = presented?[GuideDetailViewControllerSkillNameKey] as? String
        }
    }
    
    // MARK: - State Restoration
    
    let GuideDetailViewControllerTitleKey = "Title"
    let GuideDetailViewControllerSelectedIndexPathKey = "SelectedIndexPath"
    
    override func encodeRestorableState(with coder: NSCoder) {
        super.encodeRestorableState(with: coder)
        
        coder.encode(title, forKey: GuideDetailViewControllerTitleKey)
        coder.encode(sections, forKey: PropertyListBasedViewControllerSectionsKey)
        coder.encode(tableView.indexPathForSelectedRow, forKey: GuideDetailViewControllerSelectedIndexPathKey)
    }
    
    override func decodeRestorableState(with coder: NSCoder) {
        title = coder.decodeObject(forKey: GuideDetailViewControllerTitleKey) as? String
        sections = coder.decodeObject(forKey: PropertyListBasedViewControllerSectionsKey) as! NSArray
        if let selectedIndexPath = coder.decodeObject(forKey: GuideDetailViewControllerSelectedIndexPathKey) as? IndexPath {
            tableView.selectRow(at: selectedIndexPath, animated: false, scrollPosition: .top)
        }
    }
    
    // MARK: - Table View Delegate
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let sectionInfo = sections[indexPath.section] as? NSDictionary
        let rows = sectionInfo?[PropertyListBasedViewControllerRowsKey] as? NSArray
        let rowInfo = rows?[indexPath.row] as? NSDictionary
        let rowTitle = rowInfo?[PropertyListBasedViewControllerRowTitleKey] as? String
        
        if let next = rowInfo?[PropertyListBasedViewControllerNextKey] {
            let propertyListInfo: NSDictionary
            if next is String {
                let propertyListURL = Bundle.main.bundleURL.appendingPathComponent(next as! String)
                propertyListInfo = NSDictionary(contentsOf: propertyListURL)!
            } else {
                propertyListInfo = next as! NSDictionary
            }
            
            let nextSectionsInfo = propertyListInfo[PropertyListBasedViewControllerSectionsKey] as! NSArray
            
            let nextViewController = storyboard?.instantiateViewController(withIdentifier: "GuideDetailViewController") as! GuideDetailViewController
            nextViewController.title = rowTitle
            nextViewController.sections = nextSectionsInfo
            navigationController?.pushViewController(nextViewController, animated: true)
        } else if let presented = rowInfo?[PropertyListBasedViewControllerPresentedKey] as? NSDictionary {
            let presentedViewControllerName = presented[GuideDetailViewControllerViewControllerKey] as? String
            if presentedViewControllerName == "FramesPlayerViewController" {
                performSegue(withIdentifier: GuideDetailViewControllerPresentSkillMotionPlayerViewControllerSegue, sender: self)
            }
        }
    }
}

extension GuideDetailViewController: SkillMotionPlayerViewControllerDelegate {
    func willDismiss(_ skillMotionPlayerViewController: SkillMotionPlayerViewController!) {
        if UIDevice.current.userInterfaceIdiom == .pad {
            tableView.deselectRow(at: tableView.indexPathForSelectedRow!, animated: true)
        }
    }
}
